multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/variables/BatchesToProcessSerializationAdapter.java

```diff
@@ -0,0 +1,91 @@
+
+/**
+ * 
+ * Needed for backwards compatibility due to changes in how this parameter was serialized between versions - from <List<Resource> to a
+ * List<CloudServiceInstanceExtended>. This class should be removed in a following release.
+ *
+ */
+public class BatchesToProcessSerializationAdapter implements Serializer<List<List<CloudServiceInstanceExtended>>> {
+
+    public static TypeReference<List<List<CloudServiceInstanceExtended>>> BATCHES_TO_PROCESS_LIST_TYPE_REFERENCE = new TypeReference<>() {
+    };
+    public static TypeReference<List<CloudServiceInstanceExtended>> BATCHES_TO_PROCESS_ELEMENT_TYPE_REFERENCE = new TypeReference<>() {
+    };
+
+    private static TypeReference<List<Resource>> RESOURCE_LIST_TYPE_REFERENCE = new TypeReference<>() {
+    };
+
+    protected ServicesCloudModelBuilder getServicesCloudModelBuilder(VariableContainer container) {
+        CloudHandlerFactory handlerFactory = StepsUtil.getHandlerFactory((VariableScope) container);
+        DeploymentDescriptor deploymentDescriptor = VariableHandling.get(container, Variables.COMPLETE_DEPLOYMENT_DESCRIPTOR);
+        String namespace = (String) container.getVariable(Variables.MTA_NAMESPACE.getName());
+        return handlerFactory.getServicesCloudModelBuilder(deploymentDescriptor, namespace);
+    }
+
+    @Override
+    public Object serialize(List<List<CloudServiceInstanceExtended>> value) {
+        return value.stream()
+                    .map(JsonUtil::toJson)
+                    .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<List<CloudServiceInstanceExtended>> deserialize(Object serializedValue, VariableContainer container) {
+        List<List<CloudServiceInstanceExtended>> instances = new ArrayList<>();
+        if (serializedValue instanceof List<?>) {
+            List<?> outerBatchesList = (List<?>) serializedValue;
+            if (!outerBatchesList.isEmpty() && outerBatchesList.get(0) instanceof String) {
+                tryToDeserialize(outerBatchesList, container, instances);
+            }
+        }
+        return instances;
+    }
+
+    @Override
+    public List<List<CloudServiceInstanceExtended>> deserialize(Object serializedValue) {
+        return ((List<String>) serializedValue).stream()
+                                               .map(s -> JsonUtil.fromJsonBinary((s).getBytes(), BATCHES_TO_PROCESS_ELEMENT_TYPE_REFERENCE))
+                                               .collect(Collectors.toList());
+    }
+
+    private void tryToDeserialize(Object serializedValue, VariableContainer container, List<List<CloudServiceInstanceExtended>> instances) {
+        try {
+            ((List<String>) serializedValue).stream()
+                                            .map(s -> JsonUtil.fromJsonBinary((s).getBytes(), BATCHES_TO_PROCESS_ELEMENT_TYPE_REFERENCE))
+                                            .forEach(instances::add);
+        } catch (ParsingException e) {
```

Comment:
return list from this method, it is not good practice to modify collections in methods